File: src/player.ts
// Copyright (c) Zhendong Peng
// Distributed under the terms of the Modified BSD License.

import merge from 'lodash/merge'
import { DOMWidgetModel, DOMWidgetView, ISerializers } from '@jupyter-widgets/base'

import { MODULE_NAME, MODULE_VERSION } from './version'
import WaveSurfer from './wavesurfer'

// Import the CSS
import 'bootstrap/dist/css/bootstrap.min.css'

import '../css/widget.css'

export class PlayerModel extends DOMWidgetModel {
  defaults() {
    return {
      ...super.defaults(),
      _model_name: PlayerModel.model_name,
      _model_module: PlayerModel.model_module,
      _model_module_version: PlayerModel.model_module_version,
      _view_name: PlayerModel.view_name,
      _view_module: PlayerModel.view_module,
      _view_module_version: PlayerModel.view_module_version,
    }
  }

  static serializers: ISerializers = {
    ...DOMWidgetModel.serializers,
    // Add any extra serializers here
  }

  static model_name = 'PlayerModel'
  static model_module = MODULE_NAME
  static model_module_version = MODULE_VERSION
  static view_name = 'PlayerView' // Set to null if no view
  static view_module = MODULE_NAME // Set to null if no view
  static view_module_version = MODULE_VERSION
}

export class PlayerView extends DOMWidgetView {
  private _player: WaveSurfer

  render() {
    super.render()
    this.displayed.then(async () => {
      const config = {
        isStreaming: this.model.get('is_streaming'),
        language: this.model.get('language'),
      }
      this._player = WaveSurfer.create(merge({}, this.model.get('config'), config))
      this.el.appendChild(this._player.el)
      this.model.on('change:audio', () => {
        this._player.sampleRate = this.model.get('rate')
        this._player.load(this.model.get('audio'))
      })
      this.model.on('change:is_done', () => {
        this._player.setDone()
      })
    })
  }
}

File: ipyaudio/player.py
#!/usr/bin/env python
# coding: utf-8

# Copyright (c) Zhendong Peng.
# Distributed under the terms of the Modified BSD License.

import asyncio
import json
import os
import time
from pathlib import Path
from types import AsyncGeneratorType, GeneratorType
from typing import Optional, Union

import numpy as np
import torch
from IPython.display import display
from ipywidgets import DOMWidget, Label, ValueWidget, VBox, register
from lhotse import Recording
from lhotse.cut.base import Cut
from traitlets import Bool, Dict, Float, Int, Unicode

from ._frontend import module_name, module_version
from .utils import encode, merge_dicts


@register
class Player(DOMWidget, ValueWidget):
    _model_name = Unicode("PlayerModel").tag(sync=True)
    _model_module = Unicode(module_name).tag(sync=True)
    _model_module_version = Unicode(module_version).tag(sync=True)

    _view_name = Unicode("PlayerView").tag(sync=True)
    _view_module = Unicode(module_name).tag(sync=True)
    _view_module_version = Unicode(module_version).tag(sync=True)

    _audio = None
    _rate = None
    _config_path = os.path.join(os.path.dirname(__file__), "../player.json")
    config = Dict(json.load(open(_config_path))).tag(sync=True)
    audio = Unicode("").tag(sync=True)
    rate = Int(16000).tag(sync=True)
    is_streaming = Bool(False).tag(sync=True)
    is_done = Bool(False).tag(sync=True)
    language = Unicode("en").tag(sync=True)
    latency = Int(0).tag(sync=True)
    rtf = Float(0).tag(sync=True)
    verbose = Bool(False).tag(sync=True)

    def __init__(
        self,
        audio: Union[str, Path, np.ndarray, torch.Tensor, Cut, Recording, AsyncGeneratorType, GeneratorType],
        rate: Optional[int] = None,
        config: dict = {},
        language: str = "en",
        verbose: bool = False,
        **kwargs,
    ):
        super().__init__(**kwargs)
        self._audio = audio
        self._rate = rate
        self.config = merge_dicts(self.config, config)
        self.is_streaming = isinstance(audio, (AsyncGeneratorType, GeneratorType))
        self.language = language.lower()
        self.verbose = verbose
        if self.is_streaming and self.verbose:
            self.duration = 0
            self.latency_label = Label()
            self.rtf_label = Label()
            self.observe(self._update_latency_label, names="latency")
            self.observe(self._update_rtf_label, names="rtf")
            display(VBox([self, self.latency_label, self.rtf_label]))
        else:
            display(self)

    def _update_latency_label(self, change):
        label = "延迟" if self.language == "zh" else "Latency"
        self.latency_label.value = f"{label}: {self.latency} ms"

    def _update_rtf_label(self, change):
        label = "实时率" if self.language == "zh" else "Real-Time Factor"
        self.rtf_label.value = f"{label}: {self.rtf:.2f}"

    def encode_chunk(self, start, idx, chunk, rate):
        if self.is_streaming and self.verbose:
            if idx == 0:
                self.latency = int((time.time() - start) * 1000)
            self.duration += chunk.shape[1] / rate
            self.rtf = (time.time() - start) / self.duration
        self.audio, self.rate = encode(chunk, rate, False)

    async def async_encode(self, audio: AsyncGeneratorType, rate: int, start: float):
        async for idx, chunk in enumerate(audio):
            self.encode_chunk(start, idx, chunk, rate)

    def play(self):
        start = time.time()
        if isinstance(self._audio, (str, Path, np.ndarray, torch.Tensor, Cut, Recording)):
            # [num_channels, num_samples]
            self.audio, self.rate = encode(self._audio, self._rate)
        elif isinstance(self._audio, AsyncGeneratorType):
            asyncio.create_task(self.async_encode(self._audio, self._rate, start))
            self.is_done = True
        else:
            for idx, chunk in enumerate(self._audio):
                self.encode_chunk(start, idx, chunk, self._rate)
            self.is_done = True
